[WIP] Use Conda.jl if Ipython/jupyter is not found

dhoegh opened the pull request:
This is the first working version of an integrated IPython setup on all platforms.
* Make build.jl can detect a `JUPYTER` environment variable
* The Juptyer path should be output to a `const jupyter = "..."` in a IJulia/deps/deps.jl file that is generated during `Pkg.build`, and is included from IJulia.jl

stevengj commented on REQUIRE:
```diff
@@ -2,4 +2,4 @@ julia 0.3
 JSON 0.2
 ZMQ 0.1
-Compat 0.7
+Compat 0.7
```
?

stevengj commented on deps/build.jl:
```diff
@@ -2,14 +2,58 @@
+# remove deps.jl if it exists, in case build.jl fails
+isfile("deps.jl") && rm("deps.jl")
+
+if isdir(abspath(dirname(@__FILE__), "..", "..", "Conda"))
```
Just add Conda to the `REQUIRE` file and assume it is installed.

dhoegh replied:
Ok. My thought was actually to detect if Conda.jl was installed else use system python. This is also the reason why I did not add Conda.jl to `REQUIRE`. If you think it is appropiat to add Conda as dependency i will do it.

stevengj commented on deps/build.jl:
```diff
@@ -2,14 +2,58 @@
-include("jupyter.jl")
-const jupyter, jupyter_vers = find_jupyter()
+if Conda_installed
+    condapath=dirname(Conda.conda)
```
Are you sure this is right?  On Windows, the `conda` command is in the `Scripts` directory.  Shouldn't it be `joinpath(Conda.PREFIX, "bin")`?

stevengj replied:
@Luthaf mentioned in https://github.com/JuliaLang/IJulia.jl/pull/355#issuecomment-139886213 that matters are more complicated on Windows.

dhoegh replied:
It should be like this because of https://github.com/Luthaf/Conda.jl/blob/master/src/Conda.jl#L35-L36

stevengj replied:
That is the location of the `conda` script, but is it the location of the `jupyter` program?  @Luthaf's comment indicates otherwise.

Luthaf replied:
This would need the `Conda.bindir(package)` to be implemented first I think. I think the easier way to do this would be to ask callers of this function to do the executable <=> package linking by themselves.

stevengj commented on deps/build.jl:
```diff
@@ -2,14 +2,58 @@
+        joinpath(condapath,"jupyter"),convert(VersionNumber, chomp(readall(`$(joinpath(condapath,"jupyter")) kernelspec --version`)))
+    catch e1
+        try
+            joinpath(condapath,"ipython"),convert(VersionNumber, chomp(readall(`$(joinpath(condapath,"ipython")) --version`)))
```
What is the point of this?  Won't Jupyter always be installed, assuming you do `Conda.add("jupyter")`?

stevengj commented on deps/build.jl:
```diff
@@ -2,14 +2,58 @@
+if isdir(abspath(dirname(@__FILE__), "..", "..", "Conda"))
+    Conda_installed = true
+    import Conda
+    Conda.add("ipython-notebook=3.1")
```
Why are you specifying a version?

stevengj replied:
Probably we want to install `jupyter` (see JuliaLang/Interact.jl#73)

dhoegh commented:
I have updated the pull to a working condition. It do only work on master of Conda.jl. @Luthaf maybe you could tag a new version? While getting this to work I bumped into https://github.com/jupyter/notebook/issues/448. I have added a workaround for the issue.

stevengj commented on REQUIRE:
```diff
@@ -3,3 +3,4 @@ Nettle
 JSON 0.2
 ZMQ 0.1
 Compat 0.7
+Conda
```
(please end the file with a newline)

stevengj commented on deps/build.jl:
```diff
@@ -78,4 +111,11 @@ copy_config("logo-32x32.png", juliakspec)
+          const jupyter = "$(escape_string(jupyter))"
+          const jupyter_vers = $(repr(jupyter_vers))
+          """)
+end
```
newline

stevengj commented on deps/build.jl:
```diff
@@ -78,4 +111,11 @@ copy_config("logo-32x32.png", juliakspec)
 
 eprintln("Installing julia kernelspec $spec_name")
-run(`$jupyter kernelspec install --replace --user $juliakspec`)
+run(`$jupyter-kernelspec install --replace --user $juliakspec`)
```
I suppose you are doing this because of jupyter/notebook#448.

@minrk, will Jupyter always install `jupyter-kernelspec` into the same PATH as `jupyter`?  Are `jupyter-kernelspec` and `jupyter-notebook`  part of the public "API"?

I'm worried that at some point Jupyter may install the subsidiary programs like `jupyter-notebook` into a hidden internal path, or eliminate them entirely.

stevengj replied:
Also, I don't think this will work if `$jupyter` is `ipython`, since I don't think the old IPython installed separate scripts?

Maybe the best thing is:

* Try `$jupyter kernelspec --version` and `$jupyter notebook --version` to see if they work.
* If they don't work, try `$jupyter-kernelspec` and `$jupyter-notebook`
* In `deps.jl`, define a `jupyter_notebook` constant to either `["$jupyter-notebook"]` or `[jupyter, "notebook"]`, and then run it in `notebook()` via  run(`$jupyter_notebook...`) to splat the arguments

dhoegh replied:
Instead of circumventing jupyter/notebook#448 we could use IPython 3 by doing `Conda.add("ipython=3.*")` until jupyter/notebook#448 is closed.

minrk replied:
The command should definitely be `jupyter kernelspec`, but this workaround should be fine while we work out what is going on with jupyter/notebook#448. We have no plans to install entrypoints to hidden locations any time soon.

Luthaf commented:
>  @Luthaf maybe you could tag a new version?

See JuliaLang/METADATA.jl#3488

dhoegh commented:
I have pushed a new version that fixes the workaround when IPython3 is used and the new lines.

stevengj commented on deps/build.jl:
```diff
@@ -78,4 +111,15 @@ copy_config("logo-32x32.png", juliakspec)
+    # Remove the commit that added this when https://github.com/jupyter/notebook/issues/448 is closed
+    run(`$jupyter-kernelspec install --replace --user $juliakspec`)
+else
+    run(`$jupyter-kernelspec install --replace --user $juliakspec`)
```
Is this supposed to be `$jupyter kernelspec`?

dhoegh replied:
Yes it is.

stevengj commented on deps/build.jl:
```diff
@@ -1,15 +1,48 @@
+                try
+                    "ipython.bat",convert(VersionNumber, chomp(readall(`ipython.bat --version`)))
+                catch e5
+                    eprintln("No system wide Jupyter or IPython is found, got the following errors\n",
```
system wide -> system-wide
is found -> was found
, got -> ; got
errors -> errors:

stevengj commented on deps/build.jl:
```diff
@@ -1,15 +1,48 @@
+    "jupyter",convert(VersionNumber, chomp(readall(`jupyter-kernelspec --version`)))
+catch e1
+    try
+        "ipython",convert(VersionNumber, chomp(readall(`ipython --version`)))
```
I think these should check for `ipython` version 3 or later.

Maybe write:
```jl
function prog_version(prog)
    try
       return convert(VersionNumber, chomp(readall(`$prog --version`)))
    catch
       return v"0.0"
    end
end

jupyter = ""
jupyter_vers = v"0.0"
for p in ("jupyter","ipython","ipython2","ipython3","ipython.bat")
    v = prog_vers(p)
    if v >= v"3.0"
       jupyter = p
       jupyter_vers = v
       break
    end
end
if jupyter_vers < v"3.0"
    info("Could not find IPython or Jupyter 3.0 or later; installing via the Conda package.")
    Conda.add("jupyter")
    jupyter = joinpath(Conda.SCRIPTDIR,"jupyter")
    jupyter_vers = prog_vers(jupyter)
    jupyter_vers >= v"3.0" && error("failed to find $jupyter 3.0 or later")
end
```

dhoegh replied:
If we intend to use `jupyter kernelspec` in the future when https://github.com/jupyter/notebook/issues/448 is fixed we should do:
```
function prog_version(prog, arg)
    try
       return convert(VersionNumber, chomp(readall(`$prog $arg --version`)))
    catch
       return v"0.0"
    end
end

jupyter = ""
jupyter_vers = v"0.0"
for (p arg) in (("jupyter", `kernelspec`),("ipython", ``),("ipython2", ``),("ipython3", ``),("ipython.bat", ``)) # for now ("jupyter", `kernelspec`) should be ("jupyter-kernelspec", ``) due to issue
    v = prog_vers(p, arg)
    if v >= v"3.0"
       jupyter = p
       jupyter_vers = v
       break
    end
end

stevengj replied:
Both `jupyter` and `jupyter kernelspec` should return the same `--version`, so I think it should be sufficient to check for the former.

dhoegh replied:
They do not seem to provide the same version number on my system:
```
C:\Users\Hoegh>jupyter --version
4.0.4

C:\Users\Hoegh>jupyter-kernelspec --version
4.0.0
```

stevengj replied:
Hmm, same for me (MacOS with Anaconda).  @minrk, is this expected or is something wrong with Anaconda's Jupyter package?

In any case, I'm still not to worried as currently any `jupyter*` program should be fine (they all have versions > 3).

dhoegh replied:
Ok. Then I will rebase to your version of the loop without the extra argument.

stevengj commented on deps/build.jl:
```diff
@@ -78,4 +96,16 @@ copy_config("logo-32x32.png", juliakspec)
 eprintln("Installing julia kernelspec $spec_name")
-run(`$jupyter kernelspec install --replace --user $juliakspec`)
+if basename(jupyter) == "jupyter"
+    # Remove the commit that added this when https://github.com/jupyter/notebook/issues/448 is closed
```
Just say `# Just use "$jupyter kernelspec" when jupyter/notebook#448 is resolved and widely available.`  I don't think I want to bother with a separate commit for this.

stevengj commented:
Looking good.  Can you squash the commits?

stevengj commented:
Oh nevermind, I guess we can leave it as two commits.

dhoegh commented:
Thank you for merging. The readme should probably be updated and close https://github.com/JuliaLang/IJulia.jl/issues/199 and https://github.com/JuliaLang/IJulia.jl/pull/211.

stevengj commented:
Yup, updating the README now.

tkelman commented:
:clap: :fireworks:
